apollo-runtime/src/main/java/com/apollographql/apollo/internal/batch/BatchHttpCall.kt

```diff
@@ -0,0 +1,130 @@
+package com.apollographql.apollo.internal.batch
+
+import com.apollographql.apollo.api.ScalarTypeAdapters
+import com.apollographql.apollo.api.cache.http.HttpCache
+import com.apollographql.apollo.api.cache.http.HttpCachePolicy
+import com.apollographql.apollo.api.internal.Optional
+import com.apollographql.apollo.api.internal.json.BufferedSourceJsonReader
+import com.apollographql.apollo.api.internal.json.JsonWriter
+import com.apollographql.apollo.api.internal.json.ResponseJsonStreamReader
+import com.apollographql.apollo.api.internal.json.Utils
+import com.apollographql.apollo.api.internal.json.writeArray
+import com.apollographql.apollo.cache.ApolloCacheHeaders
+import com.apollographql.apollo.exception.ApolloException
+import com.apollographql.apollo.interceptor.ApolloInterceptor
+import com.apollographql.apollo.internal.interceptor.ApolloServerInterceptor
+import okhttp3.Call
+import okhttp3.Callback
+import okhttp3.HttpUrl
+import okhttp3.Request
+import okhttp3.RequestBody
+import okhttp3.Response
+import okhttp3.ResponseBody
+import okio.Buffer
+import okio.ByteString
+import java.io.IOException
+import java.nio.charset.Charset
+
+/**
+ * Represents a single HTTP call that requests data for a list of queries at once
+ */
+class BatchHttpCall(
+    private val queryList: List<QueryToBatch>,
+    private val serverUrl: HttpUrl,
+    private val httpCallFactory: Call.Factory,
+    private val scalarTypeAdapters: ScalarTypeAdapters
+) {
+
+  fun execute() {
+    val queryRequestBodyList = mutableListOf<ByteString>()
+
+    for (query in queryList) {
+      // Trigger fetch started callback
+      query.callback.onFetch(ApolloInterceptor.FetchSourceType.NETWORK)
+      // Compose request body for each query
+      queryRequestBodyList.add(
+          query.request.operation.composeRequestBody(
+              query.request.autoPersistQueries,
+              query.request.sendQueryDocument,
+              scalarTypeAdapters
+          )
+      )
+    }
+    // Combine all request bodies into one
+    val batchRequestBody = RequestBody.create(
+        ApolloServerInterceptor.MEDIA_TYPE,
+        createBatchRequestJsonBody(queryRequestBodyList)
+    )
+    // TODO assumes all queries are POST (does GET even make sense for batch HTTP calls?)
+    val requestBuilder = Request.Builder()
+        .url(serverUrl)
+        .header(ApolloServerInterceptor.HEADER_ACCEPT_TYPE, ApolloServerInterceptor.ACCEPT_TYPE)
+        .header(ApolloServerInterceptor.HEADER_CONTENT_TYPE, ApolloServerInterceptor.CONTENT_TYPE)
+        .post(batchRequestBody)
+    // For the prototype, assumes all queries have the same header
+    val firstRequest = queryList.asSequence().map { it.request }.first()
+    for (header in firstRequest.requestHeaders.headers()) {
+      val value = firstRequest.requestHeaders.headerValue(header)
+      requestBuilder.header(header, value)
+    }
+
+    // TODO add http cache headers (does it make sense for batch HTTP calls?)
```

Comment:
similarly, adding the HTTP cache headers felt weird since a batch can potentially be different between 2 runs, as the queries that will be batched is timing dependent. I guess one could get lucky and get the exact same batch (maybe for single query screens), in which case HTTP caching could come into play. Let us know what you think.

Reply:
Agreed, it doesn't make a lot of sense to have HTTP cache + query batching. Ideally we could detect that HTTP + batching is used and throw an error. Maybe based on `httpCachePolicy` ?

Reply:
And I'd say we need same kind of thing with AutoPersistedQueries too

Reply:
we actually use APQ with batching and that works really well :) 

Yeah I'll do the same for file uploads, throw an exception if used with batching with a clear message

Reply:
> we actually use APQ with batching and that works really well :)

Nice. Good to know 👍 

> Yeah I'll do the same for file uploads, throw an exception if used with batching with a clear message

File Uploads are a bit weird and work with reflexion at the moment. It might be a bit awkward to detect when it's used. If it becomes too convoluted, I think it's fine to leave that as is.

Reply:
actually, since we're only enabling batching for queries and not mutations, is file upload even possible with queries? (never used it so not sure)

Reply:
and looking at http caching, since you can make non-batched queries with the same `ApolloClient`, would be annoying to have to turn it off entirely if you still want your single queries to take advantage of it.

I'm thinking of just not adding the http cache headers for batch calls, and just silently ignore it. Not sure if I should log something to let the caller know?

Reply:
> is file upload even possible with queries?

In theory I think yes.. I don't see anything preventing someone to pass an `Upload` custom scalar as a field argument. For an example a service that could sign your files using a server-side private key:

```
query GetSignature($file: Upload) {
  signature(file: $file)
}
```

In practice, I've never seen anyone do that yet though...

> we're only enabling batching for queries and not mutations

Would it make sense to enable batching for mutations? I don't think a theoretical reason to limit to queries? 

> looking at http caching, since you can make non-batched queries with the same ApolloClient, would be annoying to have to turn it off entirely if you still want your single queries to take advantage of it.

Right , makes sense 👍 

> I'm thinking of just not adding the http cache headers for batch calls, and just silently ignore it. Not sure if I should log something to let the caller know?

Could you do something per-query? 

```kotlin
    if (batchPoller != null) {
      if (httpCachePolicy != null) {
        throw new IllegalArgumentException("Using httpCachePolicy and batching at the same time does not make sense");
      }
      interceptors.add(new ApolloBatchingInterceptor(batchPoller));
```